In [1]:
import pandas as pd
import numpy as np
import scipy.stats as st
import matplotlib.pyplot as plt
import seaborn as sns
import scipy.stats as stats

One tailed t-test - 
In a packing plant, a machine packs cartons with jars. It is supposed that a new machine will pack faster on the average than the machine currently used. To test that hypothesis, the times it takes each machine to pack ten cartons are recorded. The results, in seconds, are shown in the tables in the file files_for_lab/machine.txt. Assume that there is sufficient evidence to conduct the t test, does the data provide sufficient evidence to show if one machine is better than the other?

In [2]:
data = pd.read_csv("files_for_lab/machine.txt", encoding = "utf-16", sep = "\t", names = ["new_machine", "old_machine"], header = 0)

In [3]:
data

new_machine  old_machine
0         42.1         42.7
1         41.0         43.6
2         41.3         43.8
3         41.8         43.3
4         42.4         42.5
5         42.8         43.5
6         43.2         43.1
7         42.3         41.7
8         41.8         44.0
9         42.7         44.1

In [4]:
sample = data.sample(7)
sample_new, sample_old = sample['new_machine'], sample['old_machine']

In [5]:
sample_new_mean, sample_old_mean = sample_new.mean(), sample_old.mean()
sample_new_std, sample_old_std = sample_new.std(ddof=1), sample_old.std(ddof=1)
sp = (len(sample_new) - 1) * (sample_new_std**2) +  (len(sample_old) - 1) * (sample_old_std**2)
sp /= ( len(sample_new) + len(sample_old) - 2)
r = np.sqrt( (1/len(sample_new)) + (1/len(sample_old)) )
t = ( sample_new_mean - sample_old_mean )/ (sp * r)

print("The mean of sample new machine is {:.2f}".format(sample_new_mean))
print("The mean of sample old machine is {:.2f}".format(sample_old_mean))
print("The standard deviation of sample new machine is {:.2f}".format(sample_new_std))
print("The standard deviation of sample old machine is {:.2f}".format(sample_old_std))
print("The ratio of the sample variances is {:.2f} which is bigger than 0.5 and smaller than 2".format(sample_new_std/sample_old_std))
print("The t statistic is: {:.2f}".format(t))

The mean of sample new machine is 42.19
The mean of sample old machine is 43.10
The standard deviation of sample new machine is 0.73
The standard deviation of sample old machine is 0.77
The ratio of the sample variances is 0.94 which is bigger than 0.5 and smaller than 2
The t statistic is: -3.04


In [6]:
st.t.cdf(t,df = len(sample_new)+len(sample_old)-2)

0.005122803052231821

The negative critical value is -0.003 and the t statistic is -3.25, so we reject the null hypothesis 

--------------------------------------------------------------------------------------------------------

Matched Pairs Test - In this challenge we will compare dependent samples of data describing our Pokemon (file files_for_lab/pokemon.csv). Our goal is to see whether there is a significant difference between each Pokemon's defense and attack scores. Our hypothesis is that the defense and attack scores are equal. Compare the two columns to see if there is a statistically significant difference between them and comment your result.

In [7]:
data_1 = pd.read_csv("files_for_lab/pokemon.csv")

In [8]:
data_1.head()

#                   Name Type 1  Type 2  Total  HP  Attack  Defense  \
0  1              Bulbasaur  Grass  Poison    318  45      49       49   
1  2                Ivysaur  Grass  Poison    405  60      62       63   
2  3               Venusaur  Grass  Poison    525  80      82       83   
3  3  VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4             Charmander   Fire     NaN    309  39      52       43   

   Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0       65       65     45           1      False  
1       80       80     60           1      False  
2      100      100     80           1      False  
3      122      120     80           1      False  
4       60       50     65           1      False

In [9]:
data_1.count()

#             800
Name          800
Type 1        800
Type 2        414
Total         800
HP            800
Attack        800
Defense       800
Sp. Atk       800
Sp. Def       800
Speed         800
Generation    800
Legendary     800
dtype: int64

In [10]:
sample = data_1[["Attack","Defense"]]

In [11]:
sample["Difference"] = sample["Attack"]-sample["Defense"]
sample.head()

C:\Users\guill\AppData\Local\Temp\ipykernel_2068\712408451.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  sample["Difference"] = sample["Attack"]-sample["Defense"]


Attack  Defense  Difference
0      49       49           0
1      62       63          -1
2      82       83          -1
3     100      123         -23
4      52       43           9

In [12]:
sample_diff_mean, sample_diff_std = sample["Difference"].mean(), sample["Difference"].std(ddof=1)

t = sample_diff_mean / ( sample_diff_std / np.sqrt(sample.shape[0]) )
print("The mean of our samples differences is: {:.2f}".format(sample_diff_mean))
print("The standard deviation of our samples differences is: {:.2f}".format(sample_diff_std))
print("Our t statistics is: {:.2f}".format(t))

The mean of our samples differences is: 5.16
The standard deviation of our samples differences is: 33.73
Our t statistics is: 4.33


In [13]:
cv = st.t.ppf(1-(0.05/2),df= sample.shape[0] - 1)
cv

1.9629374611056056

Statistic equals 4.33, cv equals 1.96, so we reject H0

----------------------------------

In this activity, we will look at another example. Your task is to understand the problem and write down all the steps to set up ANOVA. After the next lesson, we will ask you to solve this problem using Python. Here are the steps that you would need to work on: - Null hypothesis - Alternate hypothesis - Level of significance - Test statistic - P-value - F table

Context
Suppose you are working as an analyst in a microprocessor chip manufacturing plant. You have been given the task of analyzing a plasma etching process with respect to changing Power (in Watts) of the plasma beam. Data was collected and provided to you to conduct statistical analysis and check if changing the power of the plasma beam has any effect on the etching rate by the machine. You will conduct ANOVA and check if there is any difference in the mean etching rate for different levels of power. You can find the data anova_lab_data.xlsx file in the files_for_lab folder

State the null hypothesis
State the alternate hypothesis
What is the significance level
What are the degrees of freedom of model, error terms, and total DoF

H0 => It stays the same; H1 => != 
We have a two sided test with a significance level of 0.05/2 on left side and 1 - 0.05/2 on rigth side


In [14]:
data_2 = pd.read_excel("files_for_lab/anova_lab_data.xlsx")
data_2

Power   Etching Rate
0   160 W          5.43
1   180 W          6.24
2   200 W          8.79
3   160 W          5.71
4   180 W          6.71
5   200 W          9.20
6   160 W          6.22
7   180 W          5.98
8   200 W          7.90
9   160 W          6.01
10  180 W          5.66
11  200 W          8.15
12  160 W          5.59
13  180 W          6.60
14  200 W          7.55

In [16]:
data_2["Power"] = data_2["Power "] 
data_2 = data_2.drop("Power ", axis = 1)

In [17]:
dg_global = 0
for w in data_2['Power'].unique():
    num_obs = len(data_2[data_2['Power'] == w])  
    dg_global  += ( ( data_2[data_2['Power'] == w]['Etching Rate'].mean() - data_2['Etching Rate'].mean() ) ** 2) * num_obs
dg_global /= (data_2['Power'].nunique() - 1)
print("The value of dg_global is {:.2f}".format(dg_global)) 

The value of dg_global is 9.09


In [18]:
dg_group = 0
for w in data_2['Power'].unique():
    for rate in data_2[data_2['Power'] == w]['Etching Rate']:
        dg_group += ( rate - data_2[data_2['Power'] == w]['Etching Rate'].mean() ) ** 2
dg_group /= ( len(data_2) - data_2['Power'].nunique() ) 

print()
print("The value of dg_group is {:.2f}".format(dg_group))


The value of dg_group is 0.25


In [19]:
F = dg_global / dg_group
print("The value of F is {:.2f}".format(F))

The value of F is 36.88


In [20]:
p_value = st.f_oneway(data_2[data_2['Power'] == '160 W']['Etching Rate'],data_2[data_2['Power'] == "180 W"]['Etching Rate'],data_2[data_2['Power'] == "200 W"]['Etching Rate'])[1]
p_value

7.506584272358903e-06

p_value is lower than 0.05 so we reject the H0 meaning, changing the power of the plasma beam has some effect on the etching rate by the machine.